### PR TITLE
FIX [`PEFT` / `Core`] Copy the state dict when passing it to `load_lora_weights`

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

As per title 

Fixes: https://github.com/huggingface/diffusers/issues/7054

There should be no reason to not copy the state dict of the lora layers if one passes a dict into `load_lora_weights`, therefore avoiding to sliently modifying the passed state_dict in-place. Added also a nice test with a state dict pushed under hf-internal-testing

cc @yiyixuxu @sayakpaul @pacman100

### CODE REVIEW

**Review by BenjaminBossan — APPROVED** [review]
In general, this LGTM, thanks for working on this.

I tried to track down the source of the mutation. If I'm not missing something, the culprits seem to be `_maybe_map_sgm_blocks_to_diffusers` and `_convert_kohya_lora_to_diffusers` [here](https://github.com/huggingface/diffusers/blob/e4b8f173b97731686e290b2eb98e7f5df2b1b322/src/diffusers/loaders/lora_conversion_utils.py) because they pop from the `state_dict`. I wonder if it wouldn't be better to create the shallow copy in these two functions. The advantage would be that if we call them from somewhere else, the `state_dict` is still not mutated. As is, we are only safe if we go via `load_lora_weights`. Right now, AFAICT, that's the only function that calls them, but this could change in the future.

File: tests/lora/test_lora_layers_peft.py [hunk 1]
```diff
@@ -1727,6 +1729,20 @@ def test_load_unload_load_kohya_lora(self):
         self.assertTrue(np.allclose(lora_images, lora_images_again, atol=1e-3))
         release_memory(pipe)
 
+    def test_empty_state_dict(self):
```
Comment (sayakpaul) [inline tests/lora/test_lora_layers_peft.py:1732]: Thank you but can we maybe add this as a fast test with smaller ckpts?

File: tests/lora/test_lora_layers_peft.py [hunk 2]
```diff
@@ -1727,6 +1729,20 @@ def test_load_unload_load_kohya_lora(self):
...
+        pipe = AutoPipelineForText2Image.from_pretrained(
+            "runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16
+        ).to("cuda")
+        pipe.scheduler = LCMScheduler.from_config(pipe.scheduler.config)
+
+        cached_file = hf_hub_download("hf-internal-testing/lcm-lora-test-sd-v1-5", "test_lora.safetensors")
+        lcm_lora = load_file(cached_file)
+
+        pipe.load_lora_weights(lcm_lora, adapter_name="lcm")
+        self.assertTrue(lcm_lora != {})
```
Comment (sayakpaul) [inline tests/lora/test_lora_layers_peft.py:1743]: WDYT of making the test more rigid by comparing the length of the state dict instead?

@BenjaminBossan what are your thoughts?

**Review by sayakpaul — COMMENTED** [review]
In general, this looks good to me. But I concur with @BenjaminBossan’s notes. Could we approach the PR along those lines? @yiyixuxu what are your thoughts?

File: tests/lora/test_lora_layers_peft.py (hunk 2 above)
Comment (BenjaminBossan) [reply tests/lora/test_lora_layers_peft.py:1743]: So basically remember the size before passing it and then ensuring that it's the same after? I don't see why not.

Comment (yiyixuxu) [reply tests/lora/test_lora_layers_peft.py:1743]: IMO we can just make another `test_load_unload_load` test but with state dict, where we can call `pipe.load_lora_weights(lcm_lora, adapter_name="lcm")` twice and make sure it gives the same results - 
that way we make sure we can re-use the state dict we passed

**Review by yiyixuxu — APPROVED** [review]
thank you!

File: tests/lora/test_lora_layers_peft.py (hunk 2 above)
Comment (younesbelkada) [reply tests/lora/test_lora_layers_peft.py:1743]: Added the test !

File: tests/lora/test_lora_layers_peft.py (hunk 1 above)
Comment (younesbelkada) [reply tests/lora/test_lora_layers_peft.py]: I think having two slow tests is fine

**Comment by younesbelkada** [comment]
Thanks everyone for the review! merging for now ! if we see any other similar issue in the future I'd be happy to refactor that a bit as suggested by Benjamin
